### DISCUSSION TITLE
Add CTE support

### DISCUSSION THREAD

**Discussion by beikov** [hibernate/hibernate-orm#5105] [Design Proposals]
Now that we have implemented support for subqueries in the from clause in 6.1 as part of https://github.com/hibernate/hibernate-orm/discussions/3768 and https://hibernate.atlassian.net/browse/HHH-3356, it is time to think about CTE support in HQL and Criteria as well.

The HQL part is pretty simple IMO. I suggest we just do what SQL does here and allow specifying a `WITH` clause for `queryExpression`. The syntax is something like this:

```
queryExpression
	: withClause? orderedQuery				# SimpleQueryGroup
	| withClause? orderedQuery (setOperator orderedQuery)+	# SetQueryGroup
	;

withClause
	: WITH cte (COMMA cte)*
	;

cte
	: identifier AS (NOT? MATERIALIZED)? LEFT_PAREN queryExpression RIGHT_PAREN searchClause? cycleClause?
	;

cteAttributes
	: identifier (COMMA identifier)*
	;

searchClause
	: SEARCH (BREADTH|DEPTH) FIRST BY searchSpecifications SET identifier
	;

searchSpecifications
	: searchSpecification (COMMA searchSpecification)*
	;

searchSpecification
	: identifier sortDirection? nullsPrecedence?
	;

cycleClause
	: CYCLE cteAttributes SET identifier (TO literal DEFAULT literal)? (USING identifier)?
	;
```

The search and cycle clause are applicable to recursive CTEs, i.e. ones that refer to the CTE itself in the second `orderedQuery` when the structure is `orderedQuery UNION ALL? orderedQuery`.

The SQM API for this is pretty simple:

```java
HibernateCriteriaBuilder cb = session.getCriteriaBuilder();

// Create the femaleContacts CTE body
JpaCriteriaQuery<Tuple> cte = cb.createTupleQuery();
JpaRoot<Contact> cteRoot = cte.from( Contact.class );
cte.multiselect( cteRoot.get( "id" ).alias( "id" ), cteRoot.get( "name" ).alias( "name" ) );
cte.where( cb.equal( cteRoot.get( "gender" ), Contact.Gender.FEMALE ) );

// Create the main query and add the femaleContacts CTE through the "with" method
final JpaCriteriaQuery<Tuple> cq = cb.createTupleQuery();
// The returned object can then be used in `from` or `join` and can be thought of like a "named subquery"
final JpaCteCriteria<Tuple> femaleContacts = cq.with( cte );

final JpaRoot<Tuple> root = cq.from( femaleContacts );

cq.multiselect( root.get( "id" ), root.get( "name" ) );
```

which would be the way to write this HQL query:

```sql
with femaleContacts as (
	select c.id id, c.name name from Contact c where c.gender = FEMALE
)
select c.id, c.name from femaleContacts c
```

My intention is to add this to the next feature release 6.2 as part of https://hibernate.atlassian.net/browse/HHH-15328. Let's move further discussions to the [PR](https://github.com/hibernate/hibernate-orm/pull/5274) that implements this.

**Comment by gavinking** [comment]
@beikov this looks sensible, but could you please add a couple of example queries showing how this would be used?

**Reply by beikov** [reply]
Sure, so here are a few example queries. Note that I tried to come up with some "real world samples", but these might be stupid. Just look at the usage and don't judge the content ;)

```sql
with unpaid_orders_older as (
  select o.id as orderId
  from Order o
  where o.paid = false
    and (year(o.date), month(o.date)) < (year(current_timestamp), month(current_timestamp))
),
unpaid_orders_this_month as (
  select o.id as orderId
  from Order o
  where o.paid = false
    and (year(o.date), month(o.date)) = (year(current_timestamp), month(current_timestamp))
)
select o.customer.id, count(*)
from Order o
join unpaid_orders_this_month um on o.id = um.orderId
where o.customer.id in (
  select o2.customer.id
  from unpaid_orders_older uo
  join Order o2 on uo.orderId = o2.id
)
group by o.customer.id
```

And here the classic sample for the recursive with clause:

```sql
with recursive children as (
  select n.id as id
  from Node n
  where n.id = :nodeToFind
  union all
  select n.id as id
  from Node n
  join children c on n.parentId = c.id
)
select n
from Node n
join children c on n.id = c.id
```

The "recursive" keyword enables the use of the currently defined CTE within the CTE. Generally, the approach to such queries is to start with an initial set of elements (the first query part of the `union all`), and refer to the currently defined CTE in the recursive part (the second query part of the `union all`) to implement this recursive/iterative processing. The recursive query part will be executed over and over again until it produces no more new tuples/rows.

If you want to know more about (recursive) CTEs, you can read into the PostgreSQL documentation: https://www.postgresql.org/docs/current/queries-with.html

**Reply by gavinking** [reply]
Thanks. So `unpaid_orders_older` is the same kind of thing as a from clause subquery, correct?

**Reply by beikov** [reply]
In that particular example, `unpaid_orders_older` and `unpaid_orders_this_month` could simply be inlined as from clause subqueries, yes. Note though, that there might be performance differences between the two representations, especially when the `materialized` keyword is used for a CTE.

>Now that we have implemented support for subqueries in the from clause in 6.1 as part of https://github.com/hibernate/hibernate-orm/discussions/3768 and https://hibernate.atlassian.net/browse/HHH-3356, it is time to think about CTE support in HQL and Criteria as well.

The messaging behind this initial sentence is, that we now have the tool (from clause subquery) to emulate non-recursive CTEs if not natively supported.

**Reply by gavinking** [reply]
What I was trying to get at is what is it in terms of the type system of HQL. I think it's just a "tuple".

**Reply by beikov** [reply]
It's a named tuple, yes

**Reply by sebersole** [reply]
> It's a named tuple, yes

Not a `TableGroup` / `TableReference`?  Some form of `ColumnReferenceQualifier`.

**Reply by gavinking** [reply]
I guess strictly-speaking it's a record type, not a tuple type; my bad.

**Reply by beikov** [reply]
>Not a TableGroup / TableReference

Ultimately, there will be a `TableGroup` / `TableReference`, but we also need some kind of `SqmPathSource` which roughly is like a named tuple/record type.
